### PR TITLE
[9.0][ADD] mrp_mto_with_stock

### PR DESCRIPTION
**Pull request by LoisRForgeFlow**
The author of this module is @michaeljohn32. I just adapt it to OCA standards, document it and do minor fixes.

MRP MTO with Stock
==================

This module extends the functionality of Manufacturing to support the creation
of procurements when there is no stock available. This allow you to pull from
stock until the quantity on hand is zero, and then create a procurement
for fulfill the MO requirements.

Configuration
=============

To configure this module, you need to:

1. Go to the products you want to follow this behaviour.
2. In the view form go to the tab *Inventory* and set the *Manufacturing
    MTO/MTS Locations*. Any other location not specified here will have the
    standard behavior.

Usage
=====

To use this module, you need to:

#. Go to *Manufacturing* and create a Manufacturing Order.
#. Click on *Confirm Production*.

### CODE REVIEW

**Comment by LoisRForgeFlow** [comment]
@michaeljohn32 Do you think this is fine?

Also if you provide me your mail I would add it to the contributors section.

**Comment by AlainVanUtterbeeck** [comment]
I did some testing and found a problem:
- we use a production zone, which is defined in the bom, so the system creates new moves on confirmation (a picklist to move the goods from stock to production). These moves are the ones that need to be verified for available and misssing stock instead of the original ones (which remain in state waiting till the picklist is issued), now the code does not handle this way of working.

**Comment by AlainVanUtterbeeck** [comment]
@lreficent 
in mrp.py, adding these 3 lines fix the problem:

    @api.one
    def action_assign(self):
        """Reserves available products to the production order but also reates
        procurements for more items if we cannot reserve enough (MTO with
        stock).
        @returns list of ids"""
        # reserve all that is available (standard behaviour):
        res = super(MrpProduction, self).action_assign()
        # try to create procurements:
        for move in self.move_lines:

+            new_move = self.env['stock.move'].search([('move_dest_id','=',move.id)])
+            if new_move:
+                move = new_move

            if (move.state == 'confirmed' and move.location_id in
                    move.product_id.mrp_mts_mto_location_ids):
                domain = [('product_id', '=', move.product_id.id),
                          ('state', '=', 'running'),
                          ('move_dest_id', '=', move.id)]

**Comment by AlainVanUtterbeeck** [comment]
@lreficent
Forget my previous remark please, it is not the solution.
I will continue testing...

**Comment by LoisRForgeFlow** [comment]
@AlainVanUtterbeeck Ok, let me know whenever you find some issue :wink:

**Comment by MiquelRForgeFlow** [comment]
I think this module should have tests.

**Comment by JordiBForgeFlow** [comment]
@mreficent @lreficent added test cases, and minor fixes, including:
- search for any procurement associated to the component and MO, regardless of the state of the existing procurement, before creating a new one.
- Do not auto-reserve the MO upon confirmation

File: mrp_mto_with_stock/models/__init__.py [hunk 1]
```diff
@@ -3,5 +3,5 @@
 # Copyright 2015 John Walsh
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from . import mrp
```
Comment (MiquelRForgeFlow) [inline mrp_mto_with_stock/models/__init__.py]: then the `mrp.py` file should be deleted

**Review by MiquelRForgeFlow — CHANGES_REQUESTED** [review]

File: mrp_mto_with_stock/models/__init__.py (hunk 1 above)
Comment (LoisRForgeFlow) [reply mrp_mto_with_stock/models/__init__.py]: fixed @mreficent

File: mrp_mto_with_stock/README.rst [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+Images
+------
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
+
+Contributors
+------------
+
+* John Walsh <?>
```
Comment (MiquelRForgeFlow) [inline mrp_mto_with_stock/README.rst]: `<John.Walsh@interclean.com>`

Comment (LoisRForgeFlow) [reply mrp_mto_with_stock/README.rst]: Thanks!

**Review by MiquelRForgeFlow — APPROVED** [review]
LGTM :+1:

**Review by florian-dacosta — APPROVED** [review]

**Comment by wangyueming** [comment]
in mrp_production.py

I think 

`                        if qty_to_procure < move.product_uom_qty:
`

should be

`                        if qty_to_procure < move.product_qty:
`

**Comment by LoisRForgeFlow** [comment]
This got merged in v10, can we move forward here too?

**Comment by LoisRForgeFlow** [comment]
@OCA/manufacturing-maintainers can we get an update on this?

**Review by jgrandguillaume — APPROVED** [review]
LGTM

**Pull request merged**